File: acwing/算法基础课/第三讲/858. Prim算法求最小生成树.py
N = 510
M = int(1e5) + 10

INF = float("inf")

g = [[INF for _ in range(N)] for _ in range(N)]
d = [INF] * N
st = [0] * N


# def prim():
#     res = 0
#     d[1] = 0
#     for j in range(n):
#         t = -1
#         for i in range(1, n+1):
#             if st[i] == 0 and (t == -1 or d[t] > d[i]):
#                 t = i

#         st[t] = 1

#         if j and d[t] == INF:
#             return "impossible"

#         res += d[t]

#         for i in range(1, n+1):
#             d[i] = min(d[i], g[t][i])

#     return res
def prim():
    res = 0
    d[1] = 0
    for j in range(n):
        t = -1
        for i in range(1, n+1):
            if st[i] == 0 and (t == -1 or d[t] > d[i]):
                t = i

        st[t] = 1

        if j and d[t] == INF:
            return "impossible"

        for i in range(1, n+1):
            if st[i] == 0:
                d[i] = min(d[i], g[t][i])
        res += d[t]
    return res


n, m = map(int, input().split())

for i in range(1, n + 1):
    g[i][i] = 0


for _ in range(m):
    a, b, c = map(int, input().split())
    g[a][b] = g[b][a] = min(g[a][b], c)

print(prim())

File: acwing/算法基础课/第三讲/858.Prim算法求最小生成树/858.Prim算法求最小生成树.cpp
#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

const int N = 505;

int g[N][N];

int st[N], d[N];

int n, m;

void prim()
{
    int res = 0;
    memset(d, 0x3f, sizeof(d));
    d[1] = 0;
    for (int i = 0; i < n; i++)
    {
        int t = -1;
        for (int j = 1; j <= n; j++)
        {
            if (!st[j] && (t == -1 || d[t] > d[j]))
                t = j;
        }
        if (d[t] == 0x3f3f3f3f)
        {
            cout << "impossible" << endl;
            return;
        }
        res += d[t];
        st[t] = 1;
        for (int j = 1; j <= n; j++)
            d[j] = min(d[j], g[t][j]);
        
    }
    cout << res << endl;
}

int main()
{
    cin >> n >> m;
    memset(g, 0x3f, sizeof(g));
    for (int i = 0; i < m; i++)
    {
        int u, v, w;
        cin >> u >> v >> w;
        g[u][v] = g[v][u] = min(g[u][v], w);
    }
    prim();
    return 0;
}